### PR TITLE
tests(app): 😌 deduplicate staking test boilerplate

### PR DESCRIPTION
**Pull request by cratelyn**
this branch adds some small tweaks to cut down on common boilerplate code found
in many of the staking component tests. this allows us to (a) easily set the
epoch duration, (b) fast forward to the end of an epoch, and (c) obtain the
identity keys of the currently defined validators. 

#### f3ecf6d5806e9b99de9fa98e0e0b85076840bef0 tests(app): 🔶 add genesis content setters
    
this is some common glue code in integration tests for the staking component.
in order to smooth the inspection of component logic that runs at epoch
boundaries, the defaults for `AppState`'s inner settings for e.g. epoch
duration are overriden to shorter values.

as is, this requres stating object literals that are somewhat verbose, even
with `..T::default()` elision. this commit adds some small `with_` helpers to
allow test cases to manipulate these settings succinctly.

#### bf8f4d14253878a38bdd1589fb15d414f4745094 tests(app): 🏃 add fast forward to next epoch helper
    
this adds an extension method that can be used to fast forward to the next
epoch. this is a very common part of test logic that exercises the staking
component.

this allows test cases to invoke an extension method to jump ahead, without
needing to import the clock component and write a loop inspecting the current
epoch after creating and executing an empty block.

#### 1c5b90a23 tests(app): 🔑 use `validator_identity_keys` accessor

this adds a `ValidatorDataRead::validator_identity_keys()` trait method.

frequently, test logic needs to keep track of validators by their identity
keys. this led to a bit of common glue code that would get the latest snapshot,
the set of all validator definitions, assert that only one validator is
present, and then take the identity key of that single validator.

now we can just get the keys, bind that single entry to a variable, or return
an error if something went wrong.

---

### checklist before requesting a review

- [x] If this code contains consensus-breaking changes, I have added the
  "consensus-breaking" label. Otherwise, I declare my belief that there are not
  consensus-breaking changes, for the following reason:

> this is only additive changes, or adjusting test code.

### CODE REVIEW

File: crates/core/app/tests/app_tracks_uptime_for_validators_only_once_active.rs [hunk 1]
```diff
@@ -73,19 +70,12 @@ async fn app_tracks_uptime_for_validators_only_once_active() -> anyhow::Result<(
...
-            [Validator { identity_key, .. }] => *identity_key,
-            unexpected => panic!("there should be one validator, got: {unexpected:?}"),
-        }
-    };
+    let [existing_validator_id] = storage
+        .latest_snapshot()
+        .validator_identity_keys()
+        .await?
+        .try_into()
+        .map_err(|keys| anyhow::anyhow!("expected one key, got: {keys:?}"))?;
```
Comment (cratelyn) [inline crates/core/app/tests/app_tracks_uptime_for_validators_only_once_active.rs:78]: an example of the benefit of being able to request a `Vec<IdentityKey>` out of the temporary storage's latest snapshot.

File: crates/core/app/tests/app_tracks_uptime_for_validators_only_once_active.rs [hunk 2]
```diff
@@ -237,19 +227,10 @@ async fn app_tracks_uptime_for_validators_only_once_active() -> anyhow::Result<(
...
-            .await?
-            .tap(|start| tracing::info!(?start, "fast forwarding to next epoch"));
-        let next = loop {
-            node.block().execute().await?;
-            let current = get_epoch().await?;
-            if current != start {
-                break current;
-            }
-        };
-        tracing::info!(?start, ?next, "finished fast forwarding to next epoch");
```
Comment (cratelyn) [inline crates/core/app/tests/app_tracks_uptime_for_validators_only_once_active.rs:252]: an example of the benefit of `fast_forward_to_next_epoch`.

File: crates/core/app/tests/common/mod.rs [hunk 3]
```diff
@@ -89,3 +89,46 @@ impl TempStorageExt for TempStorage {
...
+    ) -> Result<penumbra_sct::epoch::Epoch, anyhow::Error> {
+        use {penumbra_sct::component::clock::EpochRead, tap::Tap};
+
+        let get_epoch = || async { storage.latest_snapshot().get_current_epoch().await };
+        let start = get_epoch()
+            .await?
+            .tap(|start| tracing::info!(?start, "fast forwarding to next epoch"));
+
+        loop {
+            self.block().execute().await?;
```
Comment (erwanor) [inline crates/core/app/tests/common/mod.rs:126]: This completely works, and I think the helper method makes sense. For later, I think we will want to have a way to fast-forward to the next epoch as dynamically selected by the application. This requires only an extra look-up and gives us the capability to enshrine staking component invariant in tests (e.g. an easy one is that `Tombstoning` MUST trigger an epoch transition)

File: crates/core/component/stake/src/component/validator_handler/validator_store.rs [hunk 4]
```diff
@@ -235,6 +235,15 @@ pub trait ValidatorDataRead: StateRead {
             .try_collect()
             .await
     }
+
+    /// Returns a list of **all** known validators identity keys.
+    async fn validator_identity_keys(&self) -> Result<Vec<IdentityKey>> {
+        self.prefix(state_key::validators::definitions::prefix())
```
Comment (erwanor) [inline crates/core/component/stake/src/component/validator_handler/validator_store.rs:241]: x-ref #3846 so that it shows up on the issue

**Review by erwanor — APPROVED** [review]

File: crates/core/app/tests/common/mod.rs (hunk 3 above)
Comment (cratelyn) [reply crates/core/app/tests/common/mod.rs:126]: > For later, I think we will want to have a way to fast-forward to the next epoch as dynamically selected by the application. 

x-ref #4149, where we can address this :slightly_smiling_face:

**Pull request merged**